sonar-text-plugin/src/main/resources/org/sonar/plugins/secrets/configuration/sonarqube.yaml

```diff
@@ -0,0 +1,121 @@
+        type: Standards
+      - description: MITRE, CWE-798 - Use of Hard-coded Credentials
+        link: https://cwe.mitre.org/data/definitions/798
+        type: Standards
+      - description: MITRE, CWE-259 - Use of Hard-coded Password
+        link: https://cwe.mitre.org/data/definitions/259
+        type: Standards
+      - description: SANS Top 25 - Porous Defenses
+        link: https://www.sans.org/top25-software-errors/#cat3
+        type: Standards
+    message: Make sure this SonarQube token gets revoked, changed, and removed from the code.
+
+  detection:
+    pre:
+      include:
+        content:
+          - sonar
+          - sonarqube
+          - sonarcloud
+
+  rules:
+    - id: sonarqube-token-with-prefix
+      metadata:
+        name: SonarQube tokens should not be disclosed
+      detection:
+        matching:
+          pattern: "\\b(sq[apu]_[0-9a-f]{40})\\b"
+      examples:
+        - text: >
+            sonar.login=sqp_sonarqube_token
+          containsSecret: false
+        - text: >
+            sonar.login=sqp_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+          containsSecret: false
+        - text: >
+            sonar.projectKey=example
+            sonar.host.url=http://sonarqube.example
+            sonar.login=sqp_b4556a16fa2d28519d2451a911d2e073024010bc
+            sonar.sources=src
+            sonar.exclusions=*.tests.*
+          containsSecret: true
+          match: sqp_b4556a16fa2d28519d2451a911d2e073024010bc
+        - text: >
+            property "sonar.projectKey", "example"
+            property "sonar.organization", "example"
+            property "sonar.login", "squ_b4556a16fa2d28519d2451a911d2e073024010bc"
+            property "sonar.host.url", "https://sonarqube.example"
+          containsSecret: true
+          match: squ_b4556a16fa2d28519d2451a911d2e073024010bc
+        - text: >
+            SONARQUBE_SERVER_URL: process.env.SONARQUBE_SERVER_URL || 'https://sonarqube.example',
+            SONARQUBE_TOKEN: process.env.SONARQUBE_TOKEN || 'squ_b4556a16fa2d28519d2451a911d2e073024010bc',
+          containsSecret: true
+          match: squ_b4556a16fa2d28519d2451a911d2e073024010bc
+        - text: >
+            mvn clean verify sonar:sonar -Dsonar.login=squ_b4556a16fa2d28519d2451a911d2e073024010bc
+          containsSecret: true
+          match: squ_b4556a16fa2d28519d2451a911d2e073024010bc
+        - text: >
+            @echo off
+            setx "SONAR_TOKEN" "sqa_b4556a16fa2d28519d2451a911d2e073024010bc"
+            mvn -fn verify org.sonarsource.scanner.maven:sonar-maven-plugin:sonar
+          containsSecret: true
+          match: sqa_b4556a16fa2d28519d2451a911d2e073024010bc
+
+    - id: sonarqube-legacy-token
+      metadata:
+        name: SonarQube tokens should not be disclosed
+      detection:
+        matching:
+          pattern: "(?is)\\bD?sonar(?:qube)?[_.]?(?:login|token)\\b.{1,30}(?!sq[apu]_)\\b([0-9a-f]{40})\\b"
```

Comment:
I am not sure I get the `.{1,30}` in the middle. It appears very arbitrary to me. Moreover, the negative lookahead might bring poor performances while I don't think it's necessary.  In most of your examples, the `.` seems to only be used to match a few non-alphanum characters. Thus I would suggest simplifying that part of the expression.
 ```suggestion
          pattern: "(?is)\\bD?sonar(?:qube)?[_.]?(?:login|token)\\b\w+\\b([0-9a-f]{40})\\b"
```
It seems to still match all your examples.
Edit: beware of the space that seems to be removed in the suggestion above. No idea why. It's unintentional.

Reply:
The `.{1,30}` allows for some content to appear between the prefix and the token. For example:
```xml
property name="sonar.login" value="9acf34ff1e738bcd6d5daf102a8cfcb8a158e44e" />
```
```yaml
- name: SONAR_LOGIN
  value: 9acf34ff1e738bcd6d5daf102a8cfcb8a158e44e
```

The ideal would be to match only the token and then do a post-check on +/- 1 line for the prefix. We don't currently have the ability to set a max distance on pre-/post-check terms, so I'm using 30 characters to ensure that the prefix is reasonably close to the value.

The negative lookahead check is there to stop matching both a prefixed token and a non-prefixed token on the same value. The transition between `sq[apu]_` and the hex value also matches `\b` so some sort of lookaround check seems necessary.

Reply:
I've added a comment explaining the 30 character thing.
I've also made the 30 character match non-greedy so that it stops as soon as possible.
I've moved the `sq[apu]_` check as a negative look-behind after the token. It will therefore only execute if the rest of the regex matches and will prevent prefixed tokens from being matched.